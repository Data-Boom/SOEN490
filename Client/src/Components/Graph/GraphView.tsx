import { Box, Grid } from "@material-ui/core"
import { IGraphStateModel, newGraphState } from "../../Models/Graph/IGraphStateModel"
import React, { useState } from "react"

import Graph from './Graph'
import { GraphStateControl } from "./GraphControls/GraphStateControl"
import { IDatasetModel } from "../../Models/Datasets/IDatasetModel"
import { IGraphDatasetModel } from '../../Models/Graph/IGraphDatasetModel'
import { getGraphDatasets } from "./GraphFunctions"
import { useParams } from "react-router"

interface IGraphViewParams {
  graphStateId: string
}

export default function GraphView() {
  const { graphStateId } = useParams<IGraphViewParams>()

  const [graphDatasets, setGraphDatasets] = useState<IGraphDatasetModel[]>([])
  const [graphState, setGraphState] = useState<IGraphStateModel>({ ...newGraphState, id: graphStateId })

  const handleGraphStateChanged = (graphState: IGraphStateModel, completeDatasets: IDatasetModel[]) => {
    const graphDatasets = getGraphDatasets(completeDatasets, graphState)
    setGraphState(graphState)
    setGraphDatasets(graphDatasets)
  }

  return (
    <>
      <Box ml={8}>
        <Grid container spacing={3}>
          <Grid item container sm={7} >
            <Graph
              datasets={graphDatasets}
              axes={graphState.axes}
            />
          </Grid>
          <Grid item sm={5}>
            <Box ml={5} mr={5} mt={5}>
              <GraphStateControl
                graphState={graphState}
                onGraphStateChange={handleGraphStateChanged}
              />
            </Box>
          </Grid>
        </Grid>
      </Box >
    </>
  )
}