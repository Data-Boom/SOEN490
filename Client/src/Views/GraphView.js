import { Box, Button, Container, Grid, Paper, TextField, makeStyles } from "@material-ui/core"

import Graph from '../Components/Graph/Graph'
import React from "react"
import { useState } from "react"
import hideDatasets from '../Components/Graph/Graph'
export default function GraphView() {

  //sample datasets to try, just needs to gather from the backend instead.
  //Datalist is the list fed to the graphCreation
    //< Button onClick={hideDatasets()} > Why am I here ?</Button >
    
  const [datalist, setDatalist] = useState([])
  const [colourslist, setColourslist] = useState([])
  const [IDList, setIDList] = useState([])

  const handleRequest = () => {
    var dataset1 = [{ x: 5.1, y: 3.5 }, { x: 4.9, y: 3 }, { x: 4.7, y: 3.2 }, { x: 4.6, y: 3.1 }, { x: 5, y: 3.6 }, { x: 5.4, y: 3.9 }]
    var dataset2 = [{ x: 3, y: 8 }, { x: 9, y: 7 }, { x: 1, y: 3 }, { x: 2, y: 4 }, { x: 8, y: 1 }]
    var dataset3 = [{ x: 7, y: 6 }, { x: 2, y: 5 }, { x: 7, y: 9 }, { x: 4, y: 1 }, { x: 6, y: 7 }]
    var dataset4 = [{ x: 3, y: 7 }, { x: 1, y: 9 }, { x: 8, y: 7 }, { x: 1, y: 4 }, { x: 8, y: 5 }]
    const datalistDemo = []
    datalistDemo.push(dataset1)
    datalistDemo.push(dataset2)
    datalistDemo.push(dataset3)
    datalistDemo.push(dataset4)
    console.log(datalistDemo)
    setDatalist(datalistDemo)
    setIDList(["1" , "2" , "3" , "4"])
    setColourslist(["#3632ff", "#f20b34", "#7af684", "#000000"])
    loadOptionsList(datalistDemo)
  }

  const deleteDataset = () => {

    if (datalist.length != 0) {
      var x = document.getElementById("datasets").value;
      const datalistDemo = []
      const colourlistDemo = []
      for (var i = 0; i < datalist.length; i++) {
        if (i != x) {
          datalistDemo.push(datalist[i])
          colourlistDemo.push(colourslist[i])
        }
      }
      colourlistDemo.push(colourslist[x])
      setColourslist(colourlistDemo)
      setDatalist(datalistDemo)
      loadOptionsList(datalistDemo)
    }

  }

  const loadOptionsList = (d) => {
    var text = "<label for=\"cars\">Choose a Dataset to Delete:</label><br><select name=\"dataset\" id=\"datasets\">"
    for (var i = 0; i < d.length; i++) {
      text += "<option value=" + i + ">Dataset" + i + "</option><br>";
    }
    text += "</select><br>";
    document.getElementById("options").innerHTML = text;
  }
    
  return (
    <>
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></link>
      <h2>GraphView</h2>
          <Button onClick={handleRequest} color="primary">Load random dataset button</Button>
      <Box ml={8}>
        <Grid container spacing={3}>
          <Grid item container sm={6} >
            <Paper elevation={3}>
              <Graph
                              outerHeight={500}
                              outerWidth={768}
                              datalist={datalist}
                              colourslist={colourslist}
             IDList={IDList}
              />
            </Paper>
          </Grid>
          <Grid item sm={6}>
            <TextField id="standard-basic" label="Standard" />
            <TextField id="filled-basic" label="Filled" variant="filled" />
            <TextField id="outlined-basic" label="Outlined" variant="outlined" />
          </Grid>
        </Grid>
        <form>
          <br></br>
          <div id="options">
          </div>
        </form>
        <Button onClick={deleteDataset}>Delete dataset</Button>

      </Box>
    </>
  );
    
}
//  <Button onClick={toggleDataPoints(0)}> test</Button>