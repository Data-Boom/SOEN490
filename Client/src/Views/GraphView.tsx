// todo this highlights as an error but i dont know what to do with it, 
// it doesnt break anything, saving image as png works well

import * as svg from 'save-svg-as-png'

import { Box, Button, Grid, Modal, Paper, makeStyles } from "@material-ui/core"
import React, { useState } from "react"

import { DatasetsList } from "../Components/Graph/DatasetsList"
import Graph from '../Components/Graph/Graph'
import { ICompleteDatasetEntity } from "../Models/Datasets/ICompleteDatasetEntity"
import { IGraphDatasetModel } from '../Models/Datasets/IGraphDatasetModel'
import SearchView from './SearchView'
import { exampleExportDatasetModel } from '../Models/Datasets/IExportDatasetModel'
import queryString from "query-string"
import { useLocation } from 'react-router-dom'

//todo this is poorly hardcoded, we need to let user set their own colors, as well as support more than just 4 colors.
const defaultColors: string[] = ['#3632ff', '#f20b34', '#7af684', '#000000']

export default function GraphView() {
  const useStyles = makeStyles(() => ({
    modal: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    }
  }))

  const classes = useStyles()

  //Testing the query parser using quer-string for future use if needed
  console.log(queryString.parse(useLocation().search))


  //sample datasets to try, just needs to gather from the backend instead.
  //Datalist is the list fed to the graphCreation
  const [completeDatasets, setCompleteDatasets] = useState<ICompleteDatasetEntity[]>([])
  const [openModal, setOpenModal] = useState(false)

  const handleOpen = () => {
    setOpenModal(true)
  }

  const handleClose = () => {
    setOpenModal(false)
  }

  const handleSaveGraphImage = () => {
    svg.saveSvgAsPng(document.getElementById("graph"), "Databoom Graph.png", { backgroundColor: "#FFFFFF" })
  }

  const handleExportJson = () => {
    download("datasets.json", JSON.stringify(exampleExportDatasetModel, null, 4))
  }

  const toGraphDataset = (completeDataset: ICompleteDatasetEntity, color: string): IGraphDatasetModel => {
    const graphDataset: IGraphDatasetModel = {
      color: color,
      id: completeDataset.id,
      name: completeDataset.name,
      points: completeDataset.points
    }

    return graphDataset
  }



  //stolen from https://stackoverflow.com/questions/3665115/how-to-create-a-file-in-memory-for-user-to-download-but-not-through-server
  function download(filename: string, text: string) {
    let element = document.createElement('a')
    element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text))
    element.setAttribute('download', filename)

    element.style.display = 'none'
    document.body.appendChild(element)

    element.click()

    document.body.removeChild(element)
  }

  const onRemoveDataset = (datasetId: number) => {
    console.log("removed dataset")
    setCompleteDatasets(completeDatasets.filter(dataset => dataset.id !== datasetId))
  }

  const handleDatasetsSelected = (selectedDatasets: ICompleteDatasetEntity[]) => {
    console.log("just set datasets from handleDatasetsSelected")

    // selectedDatasets.filter will return those datasets that meet
    let notYetSelectedDatasets: ICompleteDatasetEntity[] = selectedDatasets.filter(selectedDataset => !isInStateAlready(selectedDataset))

    let mergedDatasets: ICompleteDatasetEntity[] = JSON.parse(JSON.stringify(completeDatasets))
    notYetSelectedDatasets.forEach(dataset => {
      mergedDatasets.push(dataset)
    })

    setCompleteDatasets(mergedDatasets)
  }

  const isInStateAlready = (dataset: ICompleteDatasetEntity) => {
    return completeDatasets.findIndex(existingDataset => existingDataset.id === dataset.id) != -1
  }

  return (
    <>
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></link>
      <h2>GraphView</h2>
      <Modal
        open={openModal}
        onClose={handleClose}
        className={classes.modal}
      >
        <Paper elevation={3}>
          <Box m={5}>
            <SearchView
              handleDatasetsSelected={handleDatasetsSelected}
            />
          </Box>
        </Paper>
      </Modal>
      <Box ml={8}>
        <Grid container spacing={3}>
          <Grid item container sm={5} >
            <Paper elevation={3}>
              <Graph
                outerHeight={500}
                outerWidth={768}
                datasets={completeDatasets.map((dataset, i) => toGraphDataset(dataset, defaultColors[i]))}
              />
            </Paper>
          </Grid>
          <Grid item sm={7}>
            <Box ml={5} mr={5} mt={5}>
              <Grid container direction='column'>
                <Grid item container spacing={3}>
                  <Grid item>
                    <Button onClick={handleOpen} color="primary" variant="contained">Add dataset To Graph</Button>
                  </Grid>
                  <Grid item>
                    <Button onClick={handleExportJson} color="primary" variant="contained">Export as json</Button>
                  </Grid>
                  <Grid item>
                    <Button onClick={handleSaveGraphImage} color="primary" variant="contained">Save Graph Image</Button>
                  </Grid>
                </Grid>
                <Grid item>
                  <DatasetsList datasets={completeDatasets} onRemoveDatasetClick={onRemoveDataset} />
                </Grid>
              </Grid>
            </Box>
          </Grid>
        </Grid>
      </Box>
    </>
  )
}