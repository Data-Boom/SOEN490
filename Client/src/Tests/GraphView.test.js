import React from 'react';
import renderer from "react-test-renderer";
import { mount, shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import GraphView from '../Views/GraphView'

configure({ adapter: new Adapter() });

describe('GraphView Component', () => {
    it("should match snapshot", () => {
        const tree = renderer.create(<GraphView />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls the handleRequest and loadOptionsList function on button click.', () => {
        const spy = jest.spyOn(console, 'log');
        const wrapper = mount(<GraphView />, { attachTo: document.body });

        const button = wrapper.find('button#handleRequest');
        expect(button.exists()).toEqual(true);
        button.simulate('click');
        expect(spy).toBeCalledWith('handleRequest Test');
        expect(spy).toBeCalledWith('loadOptionsList Test');
    });

});