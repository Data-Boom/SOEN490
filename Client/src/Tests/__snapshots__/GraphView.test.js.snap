// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphView Component should match snapshot 1`] = `
Array [
  <link
    href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"
    rel="stylesheet"
  />,
  <h2>
    GraphView
  </h2>,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
    disabled={false}
    id="handleRequest"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Load random dataset button
    </span>
  </button>,
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-sm-5"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
        >
          <svg
            height={500}
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 768 500"
            width={768}
          />
          <div>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="btn1"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Change X Scale
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="btn2"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Change Y Scale
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-7"
      />
    </div>
    <form>
      <br />
      <div
        id="options"
      />
    </form>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text"
      disabled={false}
      id="deleteDataset"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Delete dataset
      </span>
    </button>
  </div>,
]
`;
